src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,396 @@
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+/*
+ * This file handles continuous aggs watermark functions.
+ */
+
+#include <postgres.h>
+#include <access/xact.h>
+#include <fmgr.h>
+#include <miscadmin.h>
+#include <utils/acl.h>
+
+#include "ts_catalog/continuous_agg.h"
+#include "ts_catalog/continuous_aggs_watermark.h"
+#include "hypertable.h"
+
+typedef struct ContinuousAggregateWatermark
+{
+	int32 mat_hypertable_id;
+	MemoryContext mctx;
+	MemoryContextCallback cb;
+	CommandId cid;
+	int64 value;
+} ContinuousAggregateWatermark;
+
+/*
+ * Cache the watermark in the current transaction for better performance
+ * (by avoiding repeated max bucket calculations). The watermark will be
+ * reset at the end of the transaction, when the watermark function's input
+ * argument (materialized hypertable ID) changes, or when a new command is
+ * executed.
+ *
+ * One could potentially create a hashtable of watermarks keyed on materialized
+ * hypertable ID, but this is left as a future optimization since it doesn't
+ * seem to be common case that multiple continuous aggregates exist in the same
+ * query. Besides, the planner can constify calls to the watermark function
+ * during planning since the function is STABLE. Therefore, this is only a
+ * fallback if the planner needs to constify it many times (e.g., if used as
+ * an index condition on many chunks).
+ */
+static ContinuousAggregateWatermark *cagg_watermark_cache = NULL;
+
+/*
+ * Callback handler to reset the watermark after the transaction ends. This is
+ * triggered by the deletion of the associated memory context.
+ */
+static void
+cagg_watermark_reset(void *arg)
+{
+	cagg_watermark_cache = NULL;
+}
+
+/*
+ * ContinuousAggregateWatermark is valid for the duration of one command execution on the same
+ * materialized hypertable.
+ */
+static bool
+cagg_watermark_is_valid(const ContinuousAggregateWatermark *cagg_watermark, int32 mat_hypertable_id)
+{
+	return cagg_watermark != NULL && cagg_watermark->mat_hypertable_id == mat_hypertable_id &&
+		   cagg_watermark->cid == GetCurrentCommandId(false);
+}
+
+static void
+cagg_watermark_init_scan_by_mat_hypertable_id(ScanIterator *iterator, const int32 mat_hypertable_id)
+{
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(),
+											CONTINUOUS_AGGS_WATERMARK,
+											CONTINUOUS_AGGS_WATERMARK_PKEY);
+
+	ts_scan_iterator_scan_key_init(iterator,
+								   Anum_continuous_aggs_watermark_pkey_mat_hypertable_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(mat_hypertable_id));
+}
+
+static int64
+cagg_watermark_get(Hypertable *mat_ht)
+{
+	PG_USED_FOR_ASSERTS_ONLY short count = 0;
+	Datum watermark = (Datum) 0;
+	bool value_isnull = true;
+	ScanIterator iterator =
+		ts_scan_iterator_create(CONTINUOUS_AGGS_WATERMARK, RowExclusiveLock, CurrentMemoryContext);
```

Comment:
Why RowExclusiveLock if we are just reading the watermark. Shouldn't it be AccessShareLock?

Reply:
> Hmm... what context is being used here? You might want to check that you are not using a more long-lived context. It should probably be the tuple context since this is only relevant for the duration of the expression.

ExprContext is used